pkg/ccl/backupccl/restore_job.go

```diff
@@ -221,6 +223,65 @@ func makeBackupLocalityMap(
 	return backupLocalityMap, nil
 }
 
+func filterCompletedImportSpans(
+	completedSpansSlice []jobspb.RestoreProgress_RestoreProgressFrontierEntry,
+	importSpans []execinfrapb.RestoreSpanEntry,
+) (*spanUtils.Frontier, []execinfrapb.RestoreSpanEntry, error) {
+	var frontierCreationSlice []roachpb.Span
+	var modifiedImportSpans []execinfrapb.RestoreSpanEntry
+	for _, frontierSpanEntry := range completedSpansSlice {
+		frontierCreationSlice = append(frontierCreationSlice, frontierSpanEntry.Entry)
+	}
+	checkpointingSpanFrontier, err := spanUtils.MakeFrontier(frontierCreationSlice...)
+	if err != nil {
+		return checkpointingSpanFrontier, modifiedImportSpans, err
+	}
+	for _, completedSpanEntry := range completedSpansSlice {
+		_, err = checkpointingSpanFrontier.Forward(completedSpanEntry.Entry, completedSpanEntry.Timestamp)
+		if err != nil {
+			return checkpointingSpanFrontier, modifiedImportSpans, err
+		}
+	}
+
+	for _, importSpan := range importSpans {
+		skip := false
+		checkpointingSpanFrontier.SpanEntries(importSpan.Span, func(s roachpb.Span,
+			ts hlc.Timestamp) (done spanUtils.OpResult) {
+			if ts.Equal(hlc.Timestamp{WallTime: 1}) {
+				skip = true
+				return spanUtils.ContinueMatch
+			}
+			skip = false
+			return spanUtils.StopMatch
+		})
+		if skip {
+			continue
+		}
+		modifiedImportSpans = append(modifiedImportSpans, importSpan)
+	}
+	if len(completedSpansSlice) == 0 {
```

Comment:
i'm a bit lost. Why do we need to reconstruct the span frontier in the two branches below? we already created it above, no?

Reply:
We created the frontier with completed spans above, this is for either initializing the frontier with all empty timestamps or the remaining spans with empty timestamps

Reply:
I see. If we initialize the og spanFrontier above with `ImportSpans` instead of with `completedSpans`, i think we could then delete all the code below and still be correct, no?

Reply:
yes but I believe the time complexity for skipping over spans increases since now every span exists in the frontier

Reply:
the span frontier merges adjacent spans at the same TS! I.e. if we make a frontier with a bunch of contiguous spans all init'd at ts 0, there will be only one entry in the frontier.